### PR TITLE
Metaschema XSLT Inspector

### PR DESCRIPTION
**Pull request by wendellpiez**
# Committer Notes

Working towards #72

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema-xslt/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema-xslt/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all [website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

### CODE REVIEW

**Comment by wendellpiez** [comment]
See wiki page here: https://github.com/usnistgov/metaschema-xslt/wiki/InspectorXSLT

Discussion board to come (watch for link).

File: src/schema-gen/InspectorXSLT/TESTING.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+- Current-best Inspector implementation for any metaschemas, as generated
+- Current-best XSD expressing (a subset of) the same rules as the corresponding Inspector, for a metaschema module, for testing
+- Current-best 'composed' metaschema instance for each testing metaschema. i.e. a fully assembled and linked metaschema module, useful for debugging
+
+Within these artifacts, initial comments should give information regarding date of creation.
+
+### "Computer Model" metaschema
+
+This suite uses an extension of AJ Stein's 'computer metaschema' model for testing, maintained here as [testing/computer_metaschema.xml](testing/computer_metaschema.xml).
```
Comment (aj-stein-nist) [inline src/schema-gen/InspectorXSLT/TESTING.md]: Minor, non-blocking: technically it's the Metaschema project's model. Dave and others have made significant improvements there.

```suggestion
This suite uses an extension of 'computer metaschema' model from [examples included in the specification's repository](https://github.com/usnistgov/metaschema/blob/develop/examples/) for testing, maintained here as [testing/computer_metaschema.xml](testing/computer_metaschema.xml).
```

File: src/schema-gen/InspectorXSLT/testing/readme.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+   - [ ] XSpec MX to HTML to Markdown? functions?
+   - [ ] Build out TINY to include constraints on insert (note-ref and term-ref)
+- [ ] Review specs with eye to contributions
+- [ ] OSCAL implementation
+   - [ ] Publishing considerations
+   - [ ] Pages / platform
+- [ ] CSX version
+   - [ ] CI/CD support for deployment from updated metaschema sources?
+- [ ] Release plans and considerations
+- [ ] Review opportunities for unit tests: reuse / pubs / pool
```
Comment (aj-stein-nist) [inline src/schema-gen/InspectorXSLT/testing/readme.md]: Minor: I wouldn't block over this, but it seems better to have issues than inline TODOs in READMEs deep in the directory structure that would get lost. I would convert them to issues in the GH repo itself.

```suggestion
```

**Review by aj-stein-nist — COMMENTED** [review]
Initial feedback, but more to follow with an approval later today or tomorrow when I experiment with it more and make sure I don't find any major problems. Looks good so far. I know we discussed README organization and separation. I would think it is better to make it in a different repo to not confuse consumers of this repo, but you are the maintainer. Not my repo! :-)

**Comment by wendellpiez** [comment]
Closes #72

**Review by aj-stein-nist — APPROVED** [review]

**Comment by wendellpiez** [comment]
Holding back pending

- Support for all paths in OSCAL (cf #91 #92)
- Support for updated `allowed-values` semantics (cf https://github.com/usnistgov/metaschema/pull/413)
- Integrating PR #94

**Review by aj-stein-nist — APPROVED** [review]
@wendellpiez, I am re-reviewing and approving at the time of this message to signal I am happy with the current version of the CI additions, now with more STDOUT and STDERR logging to move us forward. If you want me to re-review after you add relevant documentation, feel free to do so, but I already reviewed the other XSLT, XSpec, XProc, and shell script changes here as needed to advance this work.

**Comment by wendellpiez** [comment]
@aj-stein-nist please feel free to re-review changes -- mainly in TESTING.md documentation (not perfect yet but better), plus also a couple of tweaks elsewhere.

The remaining desideratum for this Issue is to have CI on a PR report failure not only for broken runs but also for successful runs reporting test failures. I have vacillated but I think this is more or less essential.

A cheap/dirty way to do this is by inspecting the outputs for key (regex) match patterns.

And there could be better options. (Maybe our testing could produce a format that Github Actions could use?) Happy for any thoughts.

**Comment by wendellpiez** [comment]
In the current 'integration tests' (smoke tests) we have failures successfully reported by at least one XSpec -

https://github.com/usnistgov/metaschema-xslt/actions/runs/7791678616/job/21248238501?pr=73#step:5:144

We have the option of making either script or XSLT runtime error out, but it makes sense to catch this gracefully, *non*? (Unless an actual reporting-XML concept takes hold.)

**Pull request merged**